Firestore/Source/Core/FSTSyncEngine.mm

```diff
@@ -289,15 +289,38 @@ - (void)transactionWithRetries:(int)retries
   });
 }
 
+// Strips out mapping changes that aren't actually changes. That is, if the document already
+// existed in the target, and is being added in the target, and this is not a reset, we can
+// skip doing the work to associate the document with the target because it has already been done.
+- (void)filterUpdatesFromChange:(FSTTargetChange *)targetChange
+                  boxedTargetID:(FSTBoxedTargetID *)targetID {
+  if ([[targetChange.mapping class] isEqual:[FSTUpdateMapping class]]) {
+    FSTUpdateMapping *update = (FSTUpdateMapping *)targetChange.mapping;
+    FSTQueryView *qv = self.queryViewsByTarget[targetID];
+    FSTView *view = qv.view;
+    FSTAssert(qv, @"Missing queryview for non-limbo query: %i", [targetID intValue]);
+    FSTDocumentKeySet *added = update.addedDocuments;
+    __block FSTDocumentKeySet *result = added;
+    [added enumerateObjectsUsingBlock:^(FSTDocumentKey *docKey, BOOL *stop) {
+      if ([view.syncedDocuments containsObject:docKey]) {
+        result = [result setByRemovingObject:docKey];
+      }
+    }];
+    update.addedDocuments = result;
```

Comment:
I'm torn about suggesting this but could we somehow make this a method on the remote event? Something with a signature maybe something like this:
```objc
- (FSTRemoteEvent *)filterUpdatesUsingViews:(NSDictionary<NSNumber *, FSTQueryView *> *)views;
```
That way:
  * We can test it directly (I'm a little uncomfortable with the lack of direct testing for this change)
  * FSTRemoteEvent wouldn't become more mutable

Note that since we don't put limbo targets in the `queryViewsByTarget` dictionary, passing that as views here will naturally avoid filtering those if you changed the assertion about the missing query view into a thing that skipped filtering that target.